redash/models.py

```diff
@@ -96,6 +106,131 @@ def get(self, query_id):
+            value = value.replace(tzinfo=pytz.utc)
+        return value
+
+
+class UserDetail(walrus.Model):
+    """
+    A walrus data model to store some user data to Redis to be
+    synced to Postgres asynchronously.
+    """
+    __database__ = walrus_db
+    __namespace__ = 'redash.user.details'
+
+    user_id = walrus.IntegerField(index=True)
+    updated_at = UTCDateTimeField(index=True, default=utcnow)
+
+    @classmethod
+    def update(cls, user_id):
+        """
+        Update the user details hash using the given redis
+        pipeline, user id, optional redis id and optional user
+        details.
+
+        The fields uid, rid and updated (timestamp) are
+        enforced and can't be overwritten.
+        """
+        # try getting the user detail with the given user ID
+        # or create one if it doesn't exist yet (e.g. when key was purged)
+        try:
+            user_detail = cls.get(cls.user_id == user_id)
+            # update the timestamp with the current time
+            user_detail.updated_at = utcnow()
+            # save to Redis
+            user_detail.save()
+        except ValueError:
+            user_detail = cls.create(
+                user_id=user_id,
+                updated_at=utcnow(),
+            )
+        return user_detail
+
+    @classmethod
+    def sync(cls, chunksize=1000):
+        """
+        Syncs user details to Postgres (to the JSON field User.details).
+        """
+        to_sync = {}
+        try:
+            for user_detail in cls.all():
+                to_sync[user_detail.user_id] = user_detail
+
+            user_ids = list(to_sync.keys())
+            if not user_ids:
+                return
+            logger.info(
+                'syncing users: %s',
+                ', '.join([str(uid) for uid in user_ids])
+            )
+            # get all SQLA users that need to be updated
+            users = User.query.filter(User.id.in_(user_ids))
+            for i, user in enumerate(users):
+                update = to_sync[user.id]
+                updated_at = getattr(update, 'updated_at', None)
+                if updated_at is None:
+                    continue
+                user.active_at = updated_at
+                # flush changes to the database after a certain
+                # number of items and extend the list of keys to
+                # stop sync in case of exceptions
+                if i % chunksize == 0:
+                    db.session.flush()
+            db.session.commit()
+        except DBAPIError:
+            # reset list of keys to stop sync
+            pass
+        finally:
+            user_ids = [str(user_id) for user_id in to_sync.keys()]
+            if user_ids:
+                logger.info(
+                    'Deleting temporary user details for users %s',
+                    ', '.join(user_ids)
+                )
+                delete_query = [
+                    UserDetail.user_id == str(user_id)
```

Comment:
Won't this be a list a booleans?

Reply:
Nah, it'll be a list of query expressions since `UserDetail.user_id` is a field instance, which subclasses the [Node class](https://github.com/coleifer/walrus/blob/e9cd555a39fd9eb97d3301311f135a6c39df8b27/walrus/query.py#L164-L195) from walrus' query module which uses operator overloading to return new expressions. That way they can be used in an OR query later below.